### PR TITLE
NLP "buckets" experiment

### PR DESCRIPTION
**Pull request by BoltzmannBrain**

### CODE REVIEW

**Comment by BoltzmannBrain** [comment]
@chetan51 please review.
This runs the "buckets" experiment, where a bucket is a group of data samples with the same classification:
1. train a model on all data samples (populating a kNN)
2. for a given bucket, use a randomly selected subset of that bucket for kNN inference, getting distances to each sample in the kNN space
3. with each inference step, we concatenate the distances such that the model "improves" as more data samples are tested
4. rank the data samples closest to farthest, ignoring the inference subset, 
5. for a given bucket, the samples not used in inference are for ranking... calculate metrics for their ranks in the sorted inference results
6. repeat for several trials, and plot the results

To run a Cio model (fast): `python projects/nlp/run_buckets_experiment.py /path/to/yale_data_buckets.csv --retina name_of_4k_retina --apiKey 4k_retina_key --trials 5`

To run an HTM model (slow): `python projects/nlp/run_buckets_experiment.py /path/to/yale_data_buckets.csv -m HTMNetwork --networkConfigPath projects/nlp/data/network_configs/buckets_tp.json --generateData --retina name_of_4k_retina --apiKey 4k_retina_key --trials 5`

File: htmresearch/frameworks/nlp/htm_runner.py [hunk 1]
```diff
@@ -297,6 +299,15 @@ def _selectWinners(self, predictions):
       for p in predictions:
         counter.update(p)
       return zip(*counter.most_common(self.numClasses))[0]
+    elif self.votingmethod == "tree":
+      import pdb; pdb.set_trace()
```
Comment (chetan51) [inline htmresearch/frameworks/nlp/htm_runner.py]: Is this intentionally left in this PR?

File: htmresearch/frameworks/nlp/runner.py [hunk 2]
```diff
@@ -212,6 +203,8 @@ def setupData(self, preprocess=False):
                                   reading in samples.
     """
     self.dataDict = readCSV(self.dataPath, numLabels=self.numClasses)
+    # from htmresearch.support.csv_helper import readNSFDataset
```
Comment (chetan51) [inline htmresearch/frameworks/nlp/runner.py]: Remove commented-out code

File: projects/nlp/data/network_configs/buckets_tp.json [hunk 3]
```diff
@@ -9,13 +9,13 @@
...
   },
-  "upRegionConfig": {
+  "tpRegionConfig": {
     "regionEnabled": true,
     "regionName": "UP",
     "regionType": "py.TemporalPoolerRegion",
     "regionParams": {
       "poolerType": "union",
-      "columnCount": 2048,
+      "columnCount": 4096,
```
Comment (chetan51) [inline projects/nlp/data/network_configs/buckets_tp.json]: Seems like 4096 is overkill?

**Comment by chetan51** [comment]
:+1:

File: htmresearch/frameworks/nlp/htm_runner.py (hunk 1 above)
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/htm_runner.py]: Sorry, looks like this and some other changes in the runners got looped into this PR by mistake.

**Comment by BoltzmannBrain** [comment]
@chetan51 thanks for the review, please take a look at the latest commits.

**Comment by chetan51** [comment]
:+1:

**Pull request merged**